plasmapy/dispersion/numerical/tests/test_hollweg_.py

```diff
@@ -0,0 +1,172 @@
+                    "T_e": 1.6e6 * u.K,
+                    "T_i": 4.0e5 * u.K,
+                    "ion": Particle("p+"),
+                },
+                PhysicsWarning,
+            ),
+            # c_s/v_A << 1 PhysicsWarning
+            (
+                {
+                    "k": 10e-8 * u.rad / u.m,
+                    "theta": 88 * u.deg,
+                    "n_i": 5 * u.cm ** -3,
+                    "B": 6.98e-8 * u.T,
+                    "T_e": 1.6e6 * u.K,
+                    "T_i": 4.0e5 * u.K,
+                    "ion": Particle("p+"),
+                },
+                PhysicsWarning,
+            ),
+            # theta nearly perpendicular PhysicsWarning
+            (
+                {
+                    "k": 10e-8 * u.rad / u.m,
+                    "theta": 84 * u.deg,
+                    "n_i": 1 * u.cm ** -3,
+                    "B": 6.98e-8 * u.T,
+                    "T_e": 1.6e6 * u.K,
+                    "T_i": 4.0e5 * u.K,
+                    "ion": Particle("p+"),
+                },
+                PhysicsWarning,
+            ),
+        ],
+    )
+    def test_warning(self, kwargs, _warning):
+        """Test scenarios that raise a `Warning`."""
+        with pytest.warns(_warning):
+            hollweg(**kwargs)
+
+    @pytest.mark.parametrize(
+        "kwargs, expected",
+        [
+            (
+                {
+                    **_kwargs_single_valued,
+                    "ion": Particle("He"),
+                    "z_mean": 2.0,
+                },
+                {**_kwargs_single_valued, "ion": Particle("He +2")},
+            ),
+            #
+            # z_mean defaults to 1
+            (
+                {**_kwargs_single_valued, "ion": Particle("He")},
+                {**_kwargs_single_valued, "ion": Particle("He+")},
+            ),
+        ],
+    )
+    def test_z_mean_override(self, kwargs, expected):
+        """Test overriding behavior of kw 'z_mean'."""
+        ws = hollweg(**kwargs)
+        ws_expected = hollweg(**expected)
+
+        for mode in ws:
+            assert np.isclose(ws[mode], ws_expected[mode], atol=1e-5, rtol=1.7e-4)
+
+    @pytest.mark.parametrize(
+        "kwargs, expected",
+        [
+            ({**_kwargs_single_valued}, {"shape": ()}),
+            (
+                {
+                    **_kwargs_single_valued,
+                    "k": [1, 2, 3] * u.rad / u.m,
+                },
+                {"shape": (3,)},
+            ),
+        ],
+    )
+    def test_return_structure(self, kwargs, expected):
+        """Test the structure of the returned values."""
+        ws = hollweg(**kwargs)
+
+        assert isinstance(ws, dict)
+        assert len({"acoustic_mode", "alfven_mode", "fast_mode"} - set(ws.keys())) == 0
+
+        for mode, val in ws.items():
+            assert isinstance(val, u.Quantity)
+            assert val.unit == u.rad / u.s
+            assert val.shape == expected["shape"]
```

Comment:
@Sjbrownian is still working on the test that actually tests values.  He copied the test structure from the `two_fluid` tests, so in the end these test will look like those.

Reply:
@rocco8773 I made a test for the graph values and they all pass with atol=1e-2. I don't have anything to verify the kx*L (x-axis graph values) or the 4 different beta values shown on the graph though. I'm unsure what the best way to do this would be, I can think of an ugly way with a bunch of if statements but I'm guessing there's a better method.

Maybe we could meet sometime to discuss? I know its getting close to the Holidays so if afterwards is best for you, that works for me.

Reply:
I've made a comment on the tolerances here: https://github.com/PlasmaPy/PlasmaPy/pull/1189#discussion_r774451097 .

I don't think you really need to check `kx*L` here (by the way, if I've got the paper correct it looks like fig2 has `ky*L`, so I'm assuming a typo here?), as long as the y-axis values fit.

I also see that the test cases you're using are point values for x (or rather `ky*L`) `= {0, 1}`. Just to float the idea: if you wanted to compare to the actual curves in a sense, you could use something like https://markummitchell.github.io/engauge-digitizer/ and parametrize over pairs (value of beta, dataframe with `kx*L` and output values). I see the code takes 1D arrays for `k`, so that could be easy to accomplish.

Reply:
Yeah so these dispersion solvers we're working on are from the Bellan2012 paper (attached here, see eqn 3) which uses kx. The Hollweg1999 paper uses ky, and Bellan summarizes results from several papers so my guess is he just changed everything to kx to be consistant. Everything is the same just a change is coordinate system.

I didn't know of enguage-digitizer, I was using WebPlotDigitizer. I will look into this, thank you for the idea!


[Bellan2012.pdf](https://github.com/PlasmaPy/PlasmaPy/files/7769515/Bellan2012.pdf)

Reply:
Oooh, okay, so I did have the wrong paper. Document, document, document that in the test docstring! :)

Reply:
I should have put the Bellan paper in the docstring, thank you for pointing this out!